extensions-core/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/CachingKubernetesPeonClient.java

```diff
@@ -0,0 +1,284 @@
+ * deletion) still contact the API server directly.
+ * </p>
+ */
+public class CachingKubernetesPeonClient extends AbstractKubernetesPeonClient
+{
+  protected static final EmittingLogger log = new EmittingLogger(CachingKubernetesPeonClient.class);
+
+  public CachingKubernetesPeonClient(
+      KubernetesClientApi clientApi,
+      String namespace,
+      String overlordNamespace,
+      boolean debugJobs,
+      ServiceEmitter emitter
+  )
+  {
+    super(clientApi, namespace, overlordNamespace, debugJobs, emitter);
+  }
+
+  public CachingKubernetesPeonClient(
+      KubernetesClientApi clientApi,
+      String namespace,
+      boolean debugJobs,
+      ServiceEmitter emitter
+  )
+  {
+    super(clientApi, namespace, "", debugJobs, emitter);
+  }
+
+  @Override
+  public JobResponse waitForPeonJobCompletion(K8sTaskId taskId, long howLong, TimeUnit unit)
+  {
+    long timeoutMs = unit.toMillis(howLong);
+    long startTime = System.currentTimeMillis();
+
+    // Give the informer 2 resync periods to see the job. if it isn't seen by then, we assume the job was canceled.
+    // This is to prevent us from waiting for entire max job runtime on a job that was canceled before it even started.
+    long jobMustBeSeenBy = startTime + (clientApi.getInformerResyncPeriodMillis() * 2);
+    boolean jobSeenInCache = false;
+
+    // Set up to watch for job changes
+    CompletableFuture<Job> jobFuture = clientApi.getEventNotifier().waitForJobChange(taskId.getK8sJobName());
+
+    // We will loop until the full timeout is reached if the job is seen in cache. If the job does not show up in the cache we will exit earlier.
+    // In this loop we first check the cache to see if our job is there and complete. This avoids missing notifications that happened before we set up the watch.
+    // If the job is not complete we wait for a notification of a job change or a timeout.
+    // If it is a timeout, we loop back to check the cache again.
+    // If it is a job change notification, we check the job state and exit if complete, or loop again if still running.
+    do {
+      try {
+        Optional<Job> maybeJob = getPeonJob(taskId.getK8sJobName());
+        if (maybeJob.isPresent()) {
+          jobSeenInCache = true;
+          Job job = maybeJob.get();
+          JobResponse currentResponse = determineJobResponse(job);
+          if (currentResponse.getPhase() != PeonPhase.RUNNING) {
+            return currentResponse;
+          } else {
+            log.debug("K8s job[%s] found in cache and is still running", taskId.getK8sJobName());
+          }
+        } else if (jobSeenInCache) {
+          // Job was in cache before, but now it's gone - it was deleted and will never complete.
+          log.warn("K8s Job[%s] was not found. It can happen if the task was canceled", taskId.getK8sJobName());
+          return new JobResponse(null, PeonPhase.FAILED);
```

Comment:
How long do successfully completed jobs remain in cache?
Can we tune it using some K8s config and should we expose that as a Druid config?